src/generate-grpc-web.ts

```diff
@@ -231,5 +252,160 @@ return new Promise((resolve, reject) => {
+          host: this.host,
+          request,
+          metadata: maybeCombinedMetadata,
+          debug: this.options.debug,
+          onMessage: (next) => {
+            observer.next(next as any);
+          },
+           onEnd: (code: %T) => {
+            if (upStreamCodes.find(upStreamCode => code === upStreamCode)) {
+              setTimeout(() => {
+                upStream();
+              }, DEFAULT_TIMEOUT_TIME);
+            }
+          },
+        });
+      });
+
+      upStream();
+    }).pipe(%T());
+`,
+            BrowserHeaders,
+            grpc,
+            Code,
+            share
+          )
+        )
+    );
+}
+
+function generateGrpcWebImplObservable(): ClassSpec {
+  const maybeMetadata = TypeNames.unionType(TypeNames.anyType('grpc.Metadata'), TypeNames.UNDEFINED);
+  const optionsParam = TypeNames.anonymousType(
+    ['transport?', TypeNames.anyType('grpc.TransportFactory')],
+    ['debug?', TypeNames.BOOLEAN],
+    ['metadata?', maybeMetadata]
+  );
+  const t = TypeNames.typeVariable('T', TypeNames.bound('UnaryMethodDefinitionish'));
+  return ClassSpec.create('GrpcWebImpl')
+    .addModifiers(Modifier.EXPORT)
+    .addProperty(PropertySpec.create('host', TypeNames.STRING).addModifiers(Modifier.PRIVATE))
+    .addProperty(PropertySpec.create('options', optionsParam).addModifiers(Modifier.PRIVATE))
+    .addInterface('Rpc')
+    .addFunction(
+      FunctionSpec.createConstructor()
+        .addParameter('host', 'string')
+        .addParameter('options', optionsParam)
+        .addStatement('this.host = host')
+        .addStatement('this.options = options')
+    )
+    .addFunction(
+      FunctionSpec.create('unary')
+        .addTypeVariable(t)
+        .addParameter('methodDesc', t)
+        .addParameter('_request', TypeNames.ANY)
+        .addParameter('metadata', maybeMetadata)
+        .returns(TypeNames.anyType('Observable@rxjs').param(TypeNames.ANY))
+        .addCodeBlock(
+          CodeBlock.empty().add(
+            `const request = { ..._request, ...methodDesc.requestType };
+            const maybeCombinedMetadata =
+    metadata && this.options.metadata
+      ? new %T({ ...this.options?.metadata.headersMap, ...metadata?.headersMap })
+      : metadata || this.options.metadata;
+return new Observable(observer => {
+  %T.unary(methodDesc, {
+      request,
+      host: this.host,
+      metadata: maybeCombinedMetadata,
+      transport: this.options.transport,
+      debug: this.options.debug,
+      onEnd: (next) => {
+        if (next.status !== 0) {
+          observer.error({
+            code: next.status,
+            message: next.statusMessage,
+          });
+        } else {
+          observer.next(next.message as any);
+          observer.complete();
+        }
+      },
+    });
+  }).pipe(%T(1));
+`,
+            BrowserHeaders,
+            grpc,
+            take
+          )
+        )
+    )
+    .addFunction(
+      FunctionSpec.create('invoke')
```

Comment:
Is this just a copy/paste of the `invoke` from the `generateGrpcWebImplPromise` method? Can we share these with some sort of `.addFunction(createInvokeMethod())`?